Sync usercase asynchronously

esoergel opened the pull request:
Saving a user can be very slow, particularly when reactivating. Kishan's test user on ICDS consistently takes several minutes to save. This moves the most expensive operation into an async task, so it won't block the UI.

Context here: https://dimagi-dev.atlassian.net/browse/ICDS-16

It also moves some stuff around and renames to be clearer. Should be straightforward to review commit-by-commit.

@mkangia @kaapstorm

esoergel committed:
Move sync_user_case stuff to separate file

esoergel committed:
Mark single-use method private

esoergel committed:
Make "domain" always be the string, and simplify

esoergel committed:
Switch from signal to task

esoergel committed:
switch test to wrapper

esoergel committed:
make function private

esoergel committed:
remove shim

esoergel commented on corehq/apps/callcenter/sync_user_case.py:
```diff
@@ -173,29 +173,29 @@ def _not_same(val1, val2):
 def sync_call_center_user_case(user):
-    domain = user.project
-    config = domain.call_center_config
-    if domain and config.enabled and config.config_is_valid():
```
the `if domain` check is unnecessary here (and by extension in `sync_usercase`) - `domain.call_center_config` happens before, which has apparently not been an issue, so we don't need the check.

mkangia replied:
i saw you added it back later. Still wondering if this is possible in non-test environment. If to confirm that this is just tests, you can check for user.project if its test environment and not otherwise.

esoergel commented on corehq/apps/callcenter/sync_user_case.py:
```diff
@@ -83,7 +83,7 @@ def _domain_has_new_fields(domain, field_names):
 
 
-def sync_user_case(commcare_user, case_type, owner_id, case=None):
+def _sync_user_case(commcare_user, case_type, owner_id, case=None):
```
I didn't like having both `sync_usercase` and `sync_user_case` floating around

esoergel replied:
I don't think it's that important, I'm happy with the way it is now.  The main thing I wanted to do was make this a private method and constrain its use to this file.

esoergel committed:
fix paths

esoergel committed:
Also mock/test the usercase sync

esoergel committed:
Add back in checks for the domain object

While I'm still pretty sure this doesn't happen in production, it _does_
apparently happen a lot in tests (users assigned to a domain that
doesn't exist in the db).  Turns out this has been failing all over the
place, but because it's in a signal, those failures were silenced.  Yay
signals!

mkangia commented on corehq/apps/users/models.py:
```diff
@@ -1655,8 +1655,8 @@ def save(self, fire_signals=True, **params):
                                                           is_new_user=is_new_user)
-            sync_user_cases.delay(self)
             log_signal_errors(results, "Error occurred while syncing user (%s)", {'username': self.username})
+            sync_user_cases.delay(self)
```
Can we add the `user.project` check here instead?

plus,
why the move to place `sync_user_cases.delay` after logging?

mkangia replied:
I was just looking at that from code perspective where if you have two methods exiting on the first statement itself, why even call them.

>since all mobile workers do have projects

ya thats some HQ knowledge and if someone looks at the code and don't have that context, it would feel unnecessary.

esoergel replied:
> i saw you added it back later. Still wondering if this is possible in non-test environment. If to confirm that this is just tests, you can check for user.project if its test environment and not otherwise.

It's not possible in a non-test environment, because the old behavior would have failed if that case arose, and in fact it _did_ fail in tests, but that was swallowed by the signal call.  I suppose I could put a `if unit_testing and not user.project: return` in the task.  I'll do that.

mkangia reviewed (commented):
Though it's in the positive direction, one concern over the implementation, Since this adds a task each time a user is saved, this would add quite some celery tasks specially in case of bulk upload of users and this is handled by the default celery queue right now, which can get possibly clogged in case of high number of users. And since that runs the periodic tasks like celery heartbeat, we would receive false negatives of celery heartbeat. Does that sound right to you?

esoergel committed:
move to background queue

esoergel committed:
Don't pass user object in task

esoergel committed:
Only check for project when testing

esoergel committed:
soft deletion modifies base_doc, not doc_type

esoergel committed:
Pretend that there's a real domain

To ensure that the sync functions are called

mkangia reviewed (approved):
just reiterating on https://github.com/dimagi/commcare-hq/pull/22706/files#r241780388 in case you missed it. Fine to ignore if you chose not to do it.
Thanks for addressing all review points patiently.

I think we should give heads up to support in case they see something funky with background queue for next few days once this is merged.

esoergel commented on testapps/test_pillowtop/tests/test_user_pillow.py:
```diff
@@ -57,8 +56,7 @@ def test_kafka_user_pillow(self):
         # soft delete
-        user.doc_type = '{}{}'.format(user.doc_type, DELETED_SUFFIX)
-        user.save()
+        user.retire()
```
That's right.  This test was testing something that doesn't happen - the doc type was being given a `-deleted` suffix, but in actuality we put that suffix on the `base_doc`.  I figured this would be a more accurate test if we just used the actual "retire" function, rather than mimicking it.  The advantage of that is we don't need to try to keep the test in sync with the code - if someone changes `retire` in the future, this test will continue being realistic, when otherwise it could easily be overlooked.

esoergel committed:
Only save users once during bulkupload

CommCareUser.create sets a UUID, so they're always present - no need to
round-trip to the db just to get one.

esoergel committed:
Sync usercases synchronously in bulk upload

Bulk upload is already happening in a task, no need to spawn a separate
task per user to update usercases

esoergel committed:
Test usercase creation on bulkupload

esoergel commented:
@mkangia I think you'll like the commits I just added - was thinking about this a little more and this seemed like a simple way to avoid creating thousands of tasks on bulk upload.  In the process I also noticed that we were saving users twice each in bulk upload - the first time to make sure they have user IDs, though `CommCareUser.create` already sets IDs.

mkangia commented on corehq/apps/users/models.py:
```diff
@@ -1656,7 +1656,7 @@ def save(self, fire_signals=True, **params):
                                                           is_new_user=is_new_user)
             log_signal_errors(results, "Error occurred while syncing user (%s)", {'username': self.username})
-            sync_user_cases.delay(self)
+            sync_user_cases.delay(self._id)
```
nit: would get_id work here too instead of accessing the protected attribute?

snopoke commented:
@esoergel I haven't looked at this PR much but is it clear to you which part is slow?

esoergel commented:
@snopoke yeah, it's reopening a closed usercase (reactivating a deactivated user):
https://github.com/dimagi/commcare-hq/blob/ff4808dfe05c155b2592f648489407140a956b45/corehq/apps/callcenter/sync_user_case.py#L107-L108
I assume that has to rebuild the case from all forms which updated it, though I haven't dug in much beyond that.

kaapstorm commented on corehq/apps/users/models.py:
```diff
@@ -1651,9 +1651,11 @@ def save(self, fire_signals=True, **params):
             from .signals import commcare_user_post_save
             results = commcare_user_post_save.send_robust(sender='couch_user', couch_user=self,
                                                           is_new_user=is_new_user)
+            sync_user_cases.delay(self)
```
(fwiw, passing the ID is also good practice in case the instance can't be serialised when it's put on the queue. For us they're usually JsonObjects so not an issue, but potentially a gotcha.)

kaapstorm reviewed (commented):
:+1: Nice.
